### PR TITLE
fixed re-trigger job issue when public key deleted

### PR DESCRIPTION
**Pull request by rchikatw**
- fixed re-trigger job issue when the public key deleted
- formatted the logger in a proper format

### CODE REVIEW

File: controllers/storagecluster/provider_server.go [hunk 1]
```diff
@@ -448,12 +448,14 @@ func RandomString(l int) string {
...
 func getOnboardingJobObject(instance *ocsv1.StorageCluster) *batchv1.Job {
 
+	ptrToZero := int32(0)
 	return &batchv1.Job{
 		ObjectMeta: metav1.ObjectMeta{
 			Name:      onboardingJobName,
 			Namespace: instance.Namespace,
 		},
 		Spec: batchv1.JobSpec{
+			TTLSecondsAfterFinished: &ptrToZero, //Eligible to delete automatically when job completes
```
Comment (nb-ohad) [inline controllers/storagecluster/provider_server.go:458]: No need to have a local variable. 
This is a common case and I think we had a util for it. If not please add the following util (in the proper file in the utils package)

```
func ToPointer[T any](value T) *T {
    return &T
}
``` 

This will allow you to do
```suggestion
			TTLSecondsAfterFinished: utils.ToPointer(0)
```

File: onboarding/main.go [hunk 2]
```diff
@@ -70,6 +70,8 @@ func main() {
 			},
 		}
 
+		clientset.CoreV1().Secrets(operatorNamespace).Delete(context.Background(), onboardingPrivateKeySecretName, metav1.DeleteOptions{})
```
Comment (nb-ohad) [inline onboarding/main.go:73]: Why do we need this? what is the reason? and why outside the reconcile function?

**Review by nb-ohad — CHANGES_REQUESTED** [review]

File: onboarding/main.go (hunk 2 above)
Comment (nb-ohad) [reply onboarding/main.go:73]: That does not make sense. if the public key is deleted then a reconcile will happen and both keys will be regenerated. I don't see the edge case here

Comment (rchikatw) [reply onboarding/main.go:73]: Okay, let me describe it more clearly.
- cluster is up and running and it has both public & private key
- Now if the admin deletes the public key by mistake then what should happen?
- It will reconcile but when we call clientset.CoreV1().Secrets(operatorNamespace).Create on private key it will return an error saying the secret already exists.

Comment (nb-ohad) [reply onboarding/main.go:73]: I might be missing something here, but why do you want to call `clientset.CoreV1().Secrets(operatorNamespace).Create` on the secret? 

First, why are you working with clientset and not with a kubebuilder client? 
Second, in case the secret exists you should call `Update` and not `Create. The kube util library has a convenient `CreateOrUpdate` function that does exactly that!  (with the benefit of not actually calling the update if nothing changed)

Comment (rchikatw) [reply onboarding/main.go:73]: I have not used the Kubebuilder client because Leela suggested it's a small job bringing the heavyweight controller runtime is not good. Leela mentioned [here](https://github.com/red-hat-storage/ocs-operator/pull/2301#discussion_r1422415217). 
Yes, I agree instead of deleting the secret, I have to update if the secret already exists.

Comment (nb-ohad) [reply onboarding/main.go]: Ok. But regardless of the client, you will want to replicate the same pattern `CreateOrUpdate` implements

Comment (rchikatw) [reply onboarding/main.go]: I have addressed the comments, please resolve them?

File: tools/csv-merger/csv-merger.go [hunk 3]
```diff
@@ -1004,8 +1002,9 @@ func getUXBackendServerDeployment() appsv1.DeploymentSpec {
 							"-tls-key=/etc/tls/private/tls.key",
 							"-cookie-secret-file=/etc/proxy/secrets/session_secret",
 							"-openshift-service-account=ux-backend-server",
-							`-openshift-delegate-urls={"/":{"resource":"StorageCluster","namespace":"openshift-storage","verb":"create"}}`,
-							"-openshift-ca=/var/run/secrets/kubernetes.io/serviceaccount/ca.crt"},
+							`-openshift-delegate-urls={"/":{"group":"ocs.openshift.io","resource":"storageclusters","namespace":"openshift-storage","verb":"create"}}`,
+							"-openshift-ca=/var/run/secrets/kubernetes.io/serviceaccount/ca.crt",
+						},
```
Comment (jarrpa) [inline tools/csv-merger/csv-merger.go:1007]: Is this somehow related to the problem you're solving? If so please add some reasoning to your commit message.

**Review by jarrpa — COMMENTED** [review]
I made a few comments mostly asking for clarity on a few things. I'd also like to see anything that's not directly related to the problem solution (like the util function and the csv-merger changes) in a separate commit, but if that's not common practice in this part of the project I'm all right with leaving it as one commit.

File: controllers/util/util.go [hunk 4]
```diff
@@ -22,3 +22,7 @@ func DetectDuplicateInStringSlice(slice []string) bool {
 	}
 	return false
 }
+
+func ToPointer[T any](value T) *T {
+	return &value
+}
```
[the remark this answers is not in the document]
Comment (jarrpa) [reply controllers/util/util.go:28]: Note that, if you end up going forward with it, I intentionally suggested those names so our IDEs can more reliably show autocomplete options for the boolean values. :stuck_out_tongue_winking_eye: Though if you come up with something easier to use, by all means go for it.

File: controllers/storagecluster/provider_server.go [hunk 5]
```diff
@@ -454,6 +454,7 @@ func getOnboardingJobObject(instance *ocsv1.StorageCluster) *batchv1.Job {
 			Namespace: instance.Namespace,
 		},
 		Spec: batchv1.JobSpec{
+			TTLSecondsAfterFinished: util.ToPointer[int32](0), //Eligible to delete automatically when job completes
```
Comment (leelavg) [inline controllers/storagecluster/provider_server.go:457]: - move comment to line above and not inline
- TTL works for both Job Complete & Failed states, either change the comment to reflect that or just use `job finishes`

Comment (leelavg) [inline controllers/storagecluster/provider_server.go:457]: Isn't `util.ToPointer(int32(0))` more safer, like ide may throw error w/ `bool(0)` but not for `[bool](0)` 🤔. I believe compiler throws error for both ways though. Pls check, can stay as it is if it's not creating any issue.

File: onboarding/main.go [hunk 6]
```diff
@@ -10,7 +10,7 @@ import (
 	"github.com/red-hat-storage/ocs-operator/v4/controllers/util"
 	"golang.org/x/net/context"
 	corev1 "k8s.io/api/core/v1"
-	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/errors"
```
[the remark this answers is not in the document]
Comment (rchikatw) [reply onboarding/main.go:13]: There is no solid reason, I can undo this.

File: controllers/storagecluster/provider_server.go (hunk 5 above)
Comment (rchikatw) [reply controllers/storagecluster/provider_server.go:457]: i have checked `[bool](0)` compiler is not throwing error for this. We can stay with this

File: controllers/util/util.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (rchikatw) [reply controllers/util/util.go:28]: This is the one way of handling and I feel this is unified. No need to create separate functions to just have an autocomplete option in IDE. @nb-ohad Correct me if i am wrong and if you are interested in creating the function in the same way as Jose explained. Let me know

File: tools/csv-merger/csv-merger.go (hunk 3 above)
Comment (rchikatw) [reply tools/csv-merger/csv-merger.go:1007]: There is new BZ raised for this particular issue now and handled in separate PR added the description for it as well. https://github.com/red-hat-storage/ocs-operator/pull/2397

File: onboarding/main.go [hunk 7]
```diff
@@ -18,80 +18,95 @@ import (
...
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        onboardingPrivateKeySecretName,
+			Namespace:   operatorNamespace,
+			Annotations: map[string]string{"kubernetes.io/service-account.name": serviceAccountName},
+		},
+		Type: "kubernetes.io/service-account-token",
+		StringData: map[string]string{
+			"key": privatePem,
+		},
+	}
```
Comment (nb-ohad) [inline onboarding/main.go:87]: Why do we need the `annotations` and `type` fields?

**Review by nb-ohad — CHANGES_REQUESTED** [review]

File: controllers/storagecluster/storagecluster_controller.go [hunk 8]
```diff
@@ -166,13 +166,12 @@ func (r *StorageClusterReconciler) SetupWithManager(mgr ctrl.Manager) error {
...
 	}
 
-	onboardingSecretPredicates := builder.WithPredicates(
-		predicate.NewPredicateFuncs(
-			func(client client.Object) bool {
-				return client.GetName() == onboardingTicketPublicKeySecretName
-			},
-		),
-	)
+	secretPredicate := predicate.Funcs{
```
Comment (leelavg) [inline controllers/storagecluster/storagecluster_controller.go:169]: pls set other predicates to false (CreateFunc, UpdateFunc)

File: controllers/storagecluster/storagecluster_controller.go [hunk 9]
```diff
@@ -166,13 +166,12 @@ func (r *StorageClusterReconciler) SetupWithManager(mgr ctrl.Manager) error {
...
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			// Evaluates to false if the object has been confirmed deleted.
+			return !e.DeleteStateUnknown
```
Comment (leelavg) [inline controllers/storagecluster/storagecluster_controller.go:172]: I do not understand what this field is actually used for even after going through the docs and you don't want to invoke reconcile for every secret deletion (well, nothing usually deletes secrets again & again).

I'd suggest just check the name, `return client.GetName() == onboardingTicketPublicKeySecretName`

File: onboarding/main.go [hunk 10]
```diff
@@ -18,80 +18,95 @@ import (
...
+		os.Exit(1)
+	}
+
+	privateSecret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        onboardingPrivateKeySecretName,
+			Namespace:   operatorNamespace,
+			Annotations: map[string]string{"kubernetes.io/service-account.name": serviceAccountName},
+		},
+		Type: "kubernetes.io/service-account-token",
```
Comment (leelavg) [inline onboarding/main.go:83]: this is a opaque secret, remove `Type` field.

File: controllers/util/util.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (nb-ohad) [reply controllers/util/util.go]: @rchikatw Leela is correct, I am against adding constants at this point

File: onboarding/main.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (nb-ohad) [reply onboarding/main.go]: This is a common pattern for importing k8s errors. We do it in all of our repos and you can even see it in other community operators. Let's leave it as `kerrors`

[the remark this answers is not in the document]
Comment (rchikatw) [reply onboarding/main.go]: its already undone the changes, Its now kerrors only

File: controllers/storagecluster/provider_server.go [hunk 11]
```diff
@@ -447,13 +447,15 @@ func RandomString(l int) string {
...
-
+	ttlSecondsAfterFinished := int32(0)
 	return &batchv1.Job{
 		ObjectMeta: metav1.ObjectMeta{
 			Name:      onboardingJobName,
 			Namespace: instance.Namespace,
 		},
 		Spec: batchv1.JobSpec{
+			// Eligible to delete automatically when job finishes
+			TTLSecondsAfterFinished: &ttlSecondsAfterFinished,
```
Comment (nb-ohad) [inline controllers/storagecluster/provider_server.go:458]: Please use `utils.ToPointer(0)` instead

File: controllers/storagecluster/storagecluster_controller.go [hunk 12]
```diff
@@ -166,13 +166,17 @@ func (r *StorageClusterReconciler) SetupWithManager(mgr ctrl.Manager) error {
...
 	}
 
-	onboardingSecretPredicates := builder.WithPredicates(
-		predicate.NewPredicateFuncs(
-			func(client client.Object) bool {
-				return client.GetName() == onboardingTicketPublicKeySecretName
-			},
-		),
-	)
+	secretPredicates := predicate.Funcs{
```
Comment (nb-ohad) [inline controllers/storagecluster/storagecluster_controller.go:169]: No reason to have a struct that is not wrapped in a predicate. 

```suggestion
	secretPredicates := builder. WithPredicates(predicate.Funcs{
```

File: controllers/storagecluster/storagecluster_controller.go [hunk 13]
```diff
@@ -166,13 +166,17 @@ func (r *StorageClusterReconciler) SetupWithManager(mgr ctrl.Manager) error {
...
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			return e.Object.GetName() == onboardingTicketPublicKeySecretName
+		},
+		CreateFunc: func(e event.CreateEvent) bool {
+			return false
```
Comment (nb-ohad) [inline controllers/storagecluster/storagecluster_controller.go:174]: This is wrong as it will prevent create events on all watched secrets not just the onboarding public key secret

File: controllers/storagecluster/storagecluster_controller.go [hunk 14]
```diff
@@ -166,13 +166,17 @@ func (r *StorageClusterReconciler) SetupWithManager(mgr ctrl.Manager) error {
...
+		},
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			return false
```
Comment (nb-ohad) [inline controllers/storagecluster/storagecluster_controller.go:177]: This is wrong as it will prevent update events on all watched secrets not just the onboarding public key secret

File: controllers/storagecluster/storagecluster_controller.go [hunk 15]
```diff
@@ -166,13 +166,17 @@ func (r *StorageClusterReconciler) SetupWithManager(mgr ctrl.Manager) error {
...
-	onboardingSecretPredicates := builder.WithPredicates(
-		predicate.NewPredicateFuncs(
-			func(client client.Object) bool {
-				return client.GetName() == onboardingTicketPublicKeySecretName
-			},
-		),
-	)
+	secretPredicates := predicate.Funcs{
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			return e.Object.GetName() == onboardingTicketPublicKeySecretName
```
Comment (nb-ohad) [inline controllers/storagecluster/storagecluster_controller.go:171]: This is wrong as it will prevent delete events on all watched secrets not just the onboarding public key secret
You just need to return `true` here

**Review by nb-ohad — CHANGES_REQUESTED** [review]

**Comment by nb-ohad** [comment]
@rchikatw why aren't we setting a controller reference on the secrets and using an `Own` watch? 
If we don't do it we need to make sure to delete the secrets ourselves!

File: onboarding/main.go [hunk 16]
```diff
@@ -18,80 +21,131 @@ import (
...
+	privatePem := convertRsaPrivateKeyAsPemStr(privateKey)
+	publicPem, err := convertRsaPublicKeyAsPemStr(publicKey)
+
+	if err != nil {
+		klog.Errorf("failed to convert public key to pem str: %v", err)
+		os.Exit(1)
+	}
+
+	sc := ocsv1.StorageClusterList{}
+	instance, err := clientset.RESTClient().Get().AbsPath(
```
Comment (leelavg) [inline onboarding/main.go:60]: Initially we were only using kube native resources (secrets) and so was the comment that client-go is sufficient. 

It is evident now that runtime client abstraction is helpful and so you should be able to freely use it.

Comment (nb-ohad) [reply onboarding/main.go:60]: I agree, but please do not do the change now,  we can move back to runtime client in a later PR

File: onboarding/main.go
[the remark this answers is not in the document]
Comment (rchikatw) [reply onboarding/main.go:1]: It's been discussed over DM and we agreed to do it this in separate PR.

File: onboarding/main.go (hunk 16 above)
Comment (rchikatw) [reply onboarding/main.go:60]: I agree.

File: controllers/storagecluster/storagecluster_controller.go [hunk 17]
```diff
@@ -166,13 +166,17 @@ func (r *StorageClusterReconciler) SetupWithManager(mgr ctrl.Manager) error {
...
 	}
 
-	onboardingSecretPredicates := builder.WithPredicates(
-		predicate.NewPredicateFuncs(
-			func(client client.Object) bool {
-				return client.GetName() == onboardingTicketPublicKeySecretName
-			},
-		),
-	)
+	secretPredicates := builder.WithPredicates(predicate.Funcs{
```
Comment (leelavg) [inline controllers/storagecluster/storagecluster_controller.go:169]: why to use `builder.WithPredicates` here itself? better to follow exisitng pattern, this here locks the predicate to be only one iiuc. not mandatory if you aren't going to refresh.

**Review by leelavg — APPROVED** [review]

File: controllers/storagecluster/storagecluster_controller.go (hunk 17 above)
Comment (rchikatw) [reply controllers/storagecluster/storagecluster_controller.go:169]: I have added this because ohad said this here, https://github.com/red-hat-storage/ocs-operator/pull/2385#discussion_r1466285752

**Comment by nb-ohad** [comment]
> I made a few comments mostly asking for clarity on a few things. I'd also like to see anything that's not directly related to the problem solution (like the util function and the csv-merger changes) in a separate commit, but if that's not common practice in this part of the project I'm all right with leaving it as one commit.

@jarrpa Regarding the separation into separate PRs. From my point of view, a PR needs to be a complete contribution not to be dependent on some code that needs to be reviewed somewhere else.

Because of that, I would claim that a separate PR is an overkill but I do agree that it makes sense to separate these parts into different commits under the same PR, similar to what I ask when we have vendor changes.

File: controllers/storagecluster/storagecluster_controller.go (hunk 17 above)
Comment (nb-ohad) [reply controllers/storagecluster/storagecluster_controller.go:169]: @rchikatw I was wrong. the `WithPredicates` takes a list of `predicate.Funcs` structure instances with that new knowledge in mind I agree with @leelavg

File: onboarding/main.go [hunk 18]
```diff
@@ -18,80 +21,131 @@ import (
...
+			OwnerReferences: []metav1.OwnerReference{ownerRef},
+		},
+		StringData: map[string]string{
+			"key": publicPem,
+		},
+	}
 
+	_, err = clientset.CoreV1().
+		Secrets(operatorNamespace).
+		Create(context.Background(), publicSecret, metav1.CreateOptions{})
```
Comment (nb-ohad) [inline onboarding/main.go:145]: see previous comment about context

File: controllers/util/util.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (nb-ohad) [reply controllers/util/util.go:28]: @jarrpa Can we resolve this?

File: onboarding/main.go [hunk 19]
```diff
@@ -122,3 +158,25 @@ func convertRsaPublicKeyAsPemStr(publicKey *rsa.PublicKey) (string, error) {
...
 	return string(publicKeyPem), nil
 }
+
+func getStorageClusterMetadata(ctx context.Context, operatorNamespace string, clientset *kubernetes.Clientset) metav1.PartialObjectMetadata {
+	var storageClusterMetaDataList metav1.PartialObjectMetadataList
+	storageClusterGVKPath := fmt.Sprintf("/apis/%s/%s/namespaces/%s/%s", "ocs.openshift.io", "v1", operatorNamespace, "storageclusters")
+	instance, err := clientset.RESTClient().Get().AbsPath(storageClusterGVKPath).DoRaw(ctx)
+	if err != nil {
+		klog.Errorf("failed to get storage cluster metadata: %v", err.Error())
+		os.Exit(1)
```
Comment (nb-ohad) [inline onboarding/main.go:168]: Please do not use `os.Exit` inside nested functions. 
Return an error to the caller and act on it on the call site (upper-level main).

File: onboarding/main.go [hunk 20]
```diff
@@ -122,3 +158,25 @@ func convertRsaPublicKeyAsPemStr(publicKey *rsa.PublicKey) (string, error) {
...
+	}
+
+	if err := json.Unmarshal(instance, &storageClusterMetaDataList); err != nil {
+		klog.Errorf("failed to parse storage cluster metadata response: %v", err)
+		os.Exit(1)
```
Comment (nb-ohad) [inline onboarding/main.go:173]: Same as prev comment

File: onboarding/main.go [hunk 21]
```diff
@@ -122,3 +158,25 @@ func convertRsaPublicKeyAsPemStr(publicKey *rsa.PublicKey) (string, error) {
...
+	}
+
+	if len(storageClusterMetaDataList.Items) == 0 {
+		klog.Error("no storage cluster found")
+		os.Exit(1)
```
Comment (nb-ohad) [inline onboarding/main.go:178]: Same as prev comment

File: onboarding/main.go [hunk 22]
```diff
@@ -122,3 +158,25 @@ func convertRsaPublicKeyAsPemStr(publicKey *rsa.PublicKey) (string, error) {
 
 	return string(publicKeyPem), nil
 }
+
+func getStorageClusterMetadata(ctx context.Context, operatorNamespace string, clientset *kubernetes.Clientset) metav1.PartialObjectMetadata {
+	var storageClusterMetaDataList metav1.PartialObjectMetadataList
+	storageClusterGVKPath := fmt.Sprintf("/apis/%s/%s/namespaces/%s/%s", "ocs.openshift.io", "v1", operatorNamespace, "storageclusters")
```
Comment (nb-ohad) [inline onboarding/main.go:164]: Why are we listing all of the storage clusters? we are part of the ocs operator code base and we should have access to the constant that have the hardcoded name of the relevant storage cluster

**Review by nb-ohad — CHANGES_REQUESTED** [review]

File: onboarding/main.go
[the remark this answers is not in the document]
Comment (nb-ohad) [reply onboarding/main.go:1]: @rchikatw  We need a new package name. I agreed to move the package to another place in a separate PR but not keep the existing name

File: controllers/util/util.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (jarrpa) [reply controllers/util/util.go:28]: Sure, I'll submit a follow-up PR to fix it.

File: onboarding/main.go (hunk 22 above)
Comment (rchikatw) [reply onboarding/main.go:169]: There is no constant i could find these in the project so i have added these.

File: onboarding/main.go [hunk 23]
```diff
@@ -122,3 +163,23 @@ func convertRsaPublicKeyAsPemStr(publicKey *rsa.PublicKey) (string, error) {
...
 	return string(publicKeyPem), nil
 }
+
+func getStorageClusterMetadata(ctx context.Context, operatorNamespace string, clientset *kubernetes.Clientset) (metav1.PartialObjectMetadata, error) {
+	var storageClusterMetadataList metav1.PartialObjectMetadataList
+	storageClusterGVKPath := fmt.Sprintf("/apis/%s/%s/namespaces/%s/%s", "ocs.openshift.io", "v1", operatorNamespace, "storageclusters")
+	instance, err := clientset.RESTClient().Get().AbsPath(storageClusterGVKPath).DoRaw(ctx)
+	if err != nil {
+		klog.Errorf("failed to get storage cluster metadata: %v", err.Error())
+		return metav1.PartialObjectMetadata{}, fmt.Errorf("failed to get storage cluster metadata: %v", err)
```
Comment (nb-ohad) [inline onboarding/main.go:173]: I did not notice it in prev review but you are returning a copy of the metadata. 
There is no reason to do that, hold the metadata as a pointer and in all of the error cases return nil. 

```suggestion
		return nil, fmt.Errorf("failed to get storage cluster metadata: %v", err)
```

File: onboarding/main.go (hunk 22 above)
Comment (nb-ohad) [reply onboarding/main.go:169]: @rchikatw I don't see any local constant and you still listing all storage clusters. Am I missing something?

Comment (rchikatw) [reply onboarding/main.go:169]: Do you want me to get a specific storage cluster with name ocs-storagecluster

Comment (rchikatw) [reply onboarding/main.go]: If yes,  I have made the change to get a specific storage cluster with the name(ocs-storagecluster), I have not found this constant in ocs-operator so created a local constant.

**Comment by jarrpa** [comment]
Just to note, I poked at the whole pointer-to-value thing and effectively came to a similar conclusion to yours... with one catch: there's already a k8s library for this:

I submitted a draft PR to showcase [its usage](https://github.com/red-hat-storage/ocs-operator/pull/2428/commits/eee8fd59e714326b063ac52174a6c2f1798df955 ). I based it on this PR, so as soon as this one merges I'll mark mine as no longer a draft.

File: onboarding/main.go [hunk 24]
```diff
@@ -101,7 +142,8 @@ func newClient() (*kubernetes.Clientset, error) {
 	clientset, err := kubernetes.NewForConfig(config)
 
 	if err != nil {
-		klog.Error(err, "failed to get clientset")
+		klog.Errorf("failed to create clientset: %v", err)
+		os.Exit(1)
```
Comment (leelavg) [inline onboarding/main.go:146]: better to exit only from main and `return nil, err` here, wdyt?

Comment (rchikatw) [reply onboarding/main.go:146]: Make sense, That is how we followed, Will do the change

Comment (rchikatw) [reply onboarding/main.go]: Done.

File: controllers/storagecluster/provider_server.go [hunk 25]
```diff
@@ -454,6 +454,8 @@ func getOnboardingJobObject(instance *ocsv1.StorageCluster) *batchv1.Job {
 			Namespace: instance.Namespace,
 		},
 		Spec: batchv1.JobSpec{
+			// Eligible to delete automatically when job finishes
+			TTLSecondsAfterFinished: util.ToPointer[int32](0),
```
Comment (nb-ohad) [inline controllers/storagecluster/provider_server.go:458]: @jarrpa pointed out that k8s vendor already provides a library with similar functionality to `utils.ToPointer`:  https://pkg.go.dev/k8s.io/utils/ptr 

@rchikatw Can we please use that implementation instead and remove our util from the code base?

File: controllers/storagecluster/storagecluster_controller.go [hunk 26]
```diff
@@ -164,13 +164,17 @@ func (r *StorageClusterReconciler) SetupWithManager(mgr ctrl.Manager) error {
...
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			return true
+		},
+		CreateFunc: func(e event.CreateEvent) bool {
+			return e.Object.GetName() != onboardingTicketPublicKeySecretName
+		},
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			return e.ObjectNew.GetName() != onboardingTicketPublicKeySecretName
+		},
+	}
```
Comment (nb-ohad) [inline controllers/storagecluster/storagecluster_controller.go:177]: @rchikatw Why do we need this? 

If the operator, today or in the future, watches secrets this predicate will filter out these watches for create and update events. That is a bug!

If the underlying reason is that we want to act on events on our secret then we are already doing it with the `Owns` watch so this predicate is not needed anymore.  

If you want to NOT act on private secret events (kind of black listing these events) then you need a different predicate implementation

File: onboarding/main.go [hunk 27]
```diff
@@ -18,90 +20,128 @@ import (
...
 
 }
 
 func newClient() (*kubernetes.Clientset, error) {
 	config := runtime.GetConfigOrDie()
 	clientset, err := kubernetes.NewForConfig(config)
-
 	if err != nil {
-		klog.Error(err, "failed to get clientset")
+		return nil, fmt.Errorf("failed to create clientset: %v", err)
```
Comment (nb-ohad) [inline onboarding/main.go:144]: you are already logging this error on the call side so right now you will get:
` "failed to create clientset: failed to create clientset: <underlying err>` which is confusing. 

As we don't add new information with this message and the fact that we already log it on the call site I would suggest just returning the underlying error in this case.   

```suggestion
		return nil, err

**Comment by nb-ohad** [comment]
@rchikatw What is the reason you choose to use different formats for marshaling of private and public keys? 

You use PKCS1 for the private key:
```	privteKeyBytes := x509.MarshalPKCS1PrivateKey(privateKey)```

And PKIX for the public key:
```	publicKeyBytes, err := x509.MarshalPKIXPublicKey(publicKey)```

File: onboarding/main.go [hunk 28]
```diff
@@ -122,3 +162,18 @@ func convertRsaPublicKeyAsPemStr(publicKey *rsa.PublicKey) (string, error) {
 
 	return string(publicKeyPem), nil
 }
+
+func getStorageClusterMetadata(ctx context.Context, operatorNamespace string, clientset *kubernetes.Clientset) (*metav1.PartialObjectMetadata, error) {
+	var storageClusterMetadata metav1.PartialObjectMetadata
+	storageClusterGVKPath := fmt.Sprintf("/apis/%s/%s/namespaces/%s/%s/%s", "ocs.openshift.io", "v1", operatorNamespace, "storageclusters", storageClusterName)
```
Comment (nb-ohad) [inline onboarding/main.go:168]: 1. This is a very long line, can we please reformat it?
2. Please embed the constant part into the format string. There is no need to have them as parameters.

```suggestion
	storageClusterGVKPath := fmt.Sprintf(
		"/apis/ocs.openshift.io/v1/namespaces/%s/storageclusters/%s", 
		operatorNamespace, 
		storageClusterName,
	)
```

File: onboarding/main.go [hunk 29]
```diff
@@ -122,3 +162,18 @@ func convertRsaPublicKeyAsPemStr(publicKey *rsa.PublicKey) (string, error) {
...
+	instance, err := clientset.RESTClient().Get().AbsPath(storageClusterGVKPath).Do(ctx).Raw()
```
Comment (nb-ohad) [inline onboarding/main.go:169]: `instance` is a meaningless name, it does not convey the content or the intention behind the variable.
lease use something like `storageClusterMetadataJSON` instead.

**Review by nb-ohad — CHANGES_REQUESTED** [review]

**Comment by rchikatw** [comment]
> @rchikatw What is the reason you choose to use different formats for marshaling of private and public keys?
> 
> You use PKCS1 for the private key: ` privteKeyBytes := x509.MarshalPKCS1PrivateKey(privateKey)`
> 
> And PKIX for the public key: ` publicKeyBytes, err := x509.MarshalPKIXPublicKey(publicKey)`

A good catch actually but I never faced any issues while onboarding the application cluster. I did generate the token several times and onboarded the application cluster. But let me try once changing x509.MarshalPKIXPublicKey -> x509.MarshalPKCS1PublicKey

File: controllers/storagecluster/storagecluster_controller.go (hunk 26 above)
Comment (rchikatw) [reply controllers/storagecluster/storagecluster_controller.go:177]: So should we implement blacklisting events on private onboarding secret? or Should i use owns on all secrets like `.Owns(&corev1.Secret{}, builder.WithPredicates(predicate.GenerationChangedPredicate{})).` a default predicate

Comment (nb-ohad) [reply controllers/storagecluster/storagecluster_controller.go:177]: I think that blacklisting the private key is an overkill, we should not care if it triggers a reconcile or not as that reconcile will be a noop

**Comment by rchikatw** [comment]
> > > @rchikatw What is the reason you choose to use different formats for marshaling of private and public keys?
> > > You use PKCS1 for the private key: ` privteKeyBytes := x509.MarshalPKCS1PrivateKey(privateKey)`
> > > And PKIX for the public key: ` publicKeyBytes, err := x509.MarshalPKIXPublicKey(publicKey)`
> > 
> > 
> > A good catch actually but I never faced any issues while onboarding the application cluster. I did generate the token several times and onboarded the application cluster. But let me try once changing x509.MarshalPKIXPublicKey -> x509.MarshalPKCS1PublicKey
> > I found the reason why I am using PKIX for marshaling the public key, when a new OnboardConsumer RPC call to onboard a new OCS application cluster happens it parses the public key using the ParsePKIXPublicKey so i have used the same for the public key. refer [here](https://github.com/red-hat-storage/ocs-operator/blob/main/services/provider/server/server.go#L407) and there is no function marshell private key using PKIX
> 
> But in that case shouldn't we just align the server code to use PKCS1 as well?
> 
> P.S I just found this https://stackoverflow.com/questions/49876521/golang-x509-marshalpkixpublickey-vs-x509-marshalpkcs1publickey#:~:text=Add%20a%20comment-,1%20Answer,-Sorted%20by%3A
> 
> According to this PCIX was designed to be used for key certificates. It encodes an identity as well as encryption. I think we should align all of our code to use PKCS1. That means both private and public keys

Yes that solves the problem if we used PKCS1 at both places, Let me do that and verify

File: onboarding/main.go
[the remark this answers is not in the document]
Comment (rchikatw) [reply onboarding/main.go:1]: Yes i agree.
